crates/csharp/src/lib.rs

```diff
@@ -1198,50 +1351,195 @@ impl InterfaceGenerator<'_> {
         }
     }
 
-    fn sig_string(&mut self, func: &Function, qualifier: bool) -> String {
-        let name = func.name.to_csharp_ident();
+    fn start_resource(&mut self, id: TypeId, key: Option<&WorldKey>) {
+        let qualified = self.type_name_with_qualifier(&Type::Id(id), true);
+        let info = &self.gen.all_resources[&id];
+        let name = info.name.clone();
+        let upper_camel = name.to_upper_camel_case();
+        let docs = info.docs.clone();
+        self.print_docs(&docs);
 
-        let result_type = match func.results.len() {
-            0 => "void".into(),
-            1 => {
-                let global_prefix =
-                    self.global_if_user_type(func.results.iter_types().next().unwrap());
-                format!(
-                    "{}{}",
-                    global_prefix,
-                    self.type_name_with_qualifier(
-                        func.results.iter_types().next().unwrap(),
-                        qualifier
-                    )
-                )
+        match self.direction {
+            Direction::Import => {
+                let module_name = key
+                    .map(|key| self.resolve.name_world_key(key))
+                    .unwrap_or_else(|| "$root".into());
+
+                uwriteln!(
+                    self.src,
+                    r#"
+                    public class {upper_camel}: IDisposable {{
+                        internal int? Handle {{ get; set; }}
+
+                        internal readonly record struct THandle(int Handle);
+
+                        internal {upper_camel}(THandle handle) {{
+                            Handle = handle.Handle;
+                        }}
+
+                        public void Dispose() {{
```

Comment:
I believe you aslo want a destructor in this class that calls `Dispose(false)` .  We have no managed resources to clean up (so we do nothing with the bool), but it is the pattern.  If the user does not call `Dispose`, and the object is collected, we want to drop the imported resource.